libraries/chain/asset_evaluator.cpp

```diff
@@ -360,42 +360,127 @@ void_result asset_update_issuer_evaluator::do_apply(const asset_update_issuer_op
    return void_result();
 } FC_CAPTURE_AND_RETHROW( (o) ) }
 
-void_result asset_update_bitasset_evaluator::do_evaluate(const asset_update_bitasset_operation& o)
+void_result asset_update_bitasset_evaluator::do_evaluate(const asset_update_bitasset_operation& op)
 { try {
    database& d = db();
 
-   const asset_object& a = o.asset_to_update(d);
+   const asset_object& asset_obj = op.asset_to_update(d);
 
-   FC_ASSERT(a.is_market_issued(), "Cannot update BitAsset-specific settings on a non-BitAsset.");
+   FC_ASSERT( asset_obj.is_market_issued(), "Cannot update BitAsset-specific settings on a non-BitAsset." );
 
-   const asset_bitasset_data_object& b = a.bitasset_data(d);
-   FC_ASSERT( !b.has_settlement(), "Cannot update a bitasset after a settlement has executed" );
-   if( o.new_options.short_backing_asset != b.options.short_backing_asset )
+   FC_ASSERT( op.issuer == asset_obj.issuer, "Only asset issuer can update bitasset_data of the asset." );
+
+   const asset_bitasset_data_object& current_bitasset_data = asset_obj.bitasset_data(d);
+
+   FC_ASSERT( !current_bitasset_data.has_settlement(), "Cannot update a bitasset after a global settlement has executed" );
+
+   bool after_hf_core_922_931 = ( d.get_dynamic_global_properties().next_maintenance_time > HARDFORK_CORE_922_931_TIME );
+
+   // Are we changing the backing asset?
+   if( op.new_options.short_backing_asset != current_bitasset_data.options.short_backing_asset )
    {
-      FC_ASSERT(a.dynamic_asset_data_id(d).current_supply == 0);
-      FC_ASSERT(d.find_object(o.new_options.short_backing_asset));
+      FC_ASSERT( asset_obj.dynamic_asset_data_id(d).current_supply == 0,
+                 "Cannot update a bitasset if there is already a current supply." );
+
+      const asset_object& new_backing_asset = op.new_options.short_backing_asset(d); // check if the asset exists
 
-      if( a.issuer == GRAPHENE_COMMITTEE_ACCOUNT )
+      if( after_hf_core_922_931 ) // TODO remove this check after hard fork if things in `else` did not occur
       {
-         const asset_object& backing = a.bitasset_data(d).options.short_backing_asset(d);
-         if( backing.is_market_issued() )
+         FC_ASSERT( op.new_options.short_backing_asset != asset_obj.get_id(),
+                    "Cannot update an asset to be backed by itself." );
+
+         if( current_bitasset_data.is_prediction_market )
+         {
+            FC_ASSERT( asset_obj.precision == new_backing_asset.precision,
+                       "The precision of the asset and backing asset must be equal." );
+         }
+
+         if( asset_obj.issuer == GRAPHENE_COMMITTEE_ACCOUNT )
```

Comment:
The logic here does not match that in asset_create_operation:
https://github.com/bitshares/bitshares-core/blob/master/libraries/chain/asset_evaluator.cpp#L96-L107

Nobody is allowed to create an MPA backed by an MPA backed by an MPA. Here we only check this for committee-owned assets.

Reply:
Even if we checked here, it doesn't prevent the backing MPA which was backed by CORE or UIA from being changed to be backed by another MPA. Although technically not impossible to implement, E.G. we can maintain a variable to indicate an MPA is backing how many other MPA's, when the value is greater than zero, disallow changing it to be backed by another MPA (this approach will effectively avoid potential cycles).

However, I think the rule is debatable,
* if we've done it, another user can create an MPA backed by your MPA to prevent you from changing your MPA's backing asset to another MPA, which IMHO is some kind of attack factor;
* I image there will be potential good use cases for "an MPA backed by an MPA backed by an MPA".

Reply:
For committee-owned MPA's, there is another missing check: assume that XCD backed by USD backed by CORE (which is intended), if USD's issuer is changed to another account (which is allowed), then USD's backing asset is changed to an UIA (which is allowed as well), we'll end up with XCD owned by committee but not backing by CORE, which is unintended IMHO.

To fix this, probably better rules would be:
* don't allow changing an MPA's backing asset ID if the MPA is backing a committee-owned MPA (avoid the second change), and/or
* don't allow a committee-owned MPA to be backed by an asset which is not CORE nor owned by committee (avoid the first change)

I prefer the second rule. To get this done, we also need to modify the logic about updating an asset's issuer.

To change rules, we may need a BSIP.

Reply:
IMO the intended rules are spelled out in asset_create. asset_update_bitasset is buggy in that it only partially applies these rules (and fails to do even that properly).
IMO adapting asset_update_bitasset to fulfill the intended rules is a bugfix and does not require a BSIP.

IMO we can ignore the problems with transferring ownership from committee to users, or the committee creating MPAs backed by MPAs belonging to different users, and simply assume that the committee knows what they're doing.